### PR TITLE
ENG-12280 improvement of partition master assignment  with missing node startup

### PR DESCRIPTION
**Pull request by zeemanhe**
assign masters more evenly among the nodes during cluster startup with missing nodes.

### CODE REVIEW

**Comment by zeemanhe** [comment]
Unit test on pro: https://github.com/VoltDB/pro/pull/1881

File: src/frontend/org/voltdb/AbstractTopology.java [hunk 1]
```diff
@@ -1403,17 +1404,32 @@ public static AbstractTopology shiftPartitionLeaders(AbstractTopology topology,
                 mutableHost.partitions.add(mp);
             }
             int leaderId = partition.leaderHostId;
-            if (missingHosts.contains(partition.leaderHostId)) {
-                //find the first non-missing host that has the partition
+            if (missingHosts.contains(leaderId)) {
+                List<Host> perspectiveHosts = Lists.newArrayList();
                 for (Host host : topology.hostsById.values()) {
                     if (!missingHosts.contains(host.id)) {
                         List<Integer> list = topology.getPartitionIdList(host.id);
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/AbstractTopology.java]: maybe rename "list" with "partitionListOnNonMissingHost"?

File: src/frontend/org/voltdb/AbstractTopology.java [hunk 2]
```diff
@@ -1403,17 +1404,32 @@ public static AbstractTopology shiftPartitionLeaders(AbstractTopology topology,
...
+                            perspectiveHosts.add(host);
                         }
                     }
                 }
+
+                //Place the partition master to a node which hosts the partition and has the least masters.
+                assert(!perspectiveHosts.isEmpty());
+                perspectiveHosts.sort((Host a, Host b) -> {
+                    int result = 0;
+                    for (Partition p : topology.partitionsById.values()) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/AbstractTopology.java]: int partitionLeaderCountForHostA = 0;
for (Partition p: a.partitions) {
    if (p.leaderHostId == a.id) {
        partitionLeaderCountForHostA++;
    }
}

// do the same calculation for host B
// Compare CountA with CountB

The above calculation minimize the number of iterations on ALL partitions, maybe better to read than using one "result" to compare.

Also, do you need to sort them all to find the least partition leader contained Host?

File: src/frontend/org/voltdb/AbstractTopology.java (hunk 1 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/AbstractTopology.java]: ok

File: src/frontend/org/voltdb/AbstractTopology.java (hunk 2 above)
Comment (zeemanhe) [inline src/frontend/org/voltdb/AbstractTopology.java]: The algo has been updated.  only go through the partitions on host and find the number of leaders.  The sort is for quick access.  The perspective list is very small (<= K).

**Pull request merged**